### PR TITLE
Novitsky/account management

### PR DESCRIPTION
**Pull request by SergeyNovitsky**

### CODE REVIEW

File: OutOfSchool/OutOfSchool.IdentityServer/ViewModels/ChangeEmailViewModel.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+        [DataType(DataType.EmailAddress)]
+        [Required(ErrorMessage = "Email is required")]
+        public string Email { get; set; }
+
+        [DataType(DataType.EmailAddress)]
+        [Required(ErrorMessage = "Email is required")]
+        public string ConfirmEmail { get; set; }
+
+        public string ReturnUrl { get; set; }
```
Comment (DmyMi) [inline OutOfSchool/OutOfSchool.IdentityServer/ViewModels/ChangeEmailViewModel.cs:19]: i don't see this field used anywhere. Maybe we can use it in some kind of button in `ConfirmChange.cshtml` to redirect user back to frontend page.

File: OutOfSchool/OutOfSchool.EmailSender/ServiceProviderExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+namespace OutOfSchool.EmailSender
+{
+    public static class ServiceProviderExtensions
+    {
+        public static IServiceCollection AddEmailSender(
+            this IServiceCollection services,
+            Action<SmtpOptions> options)
+        {
+            services.AddSingleton<IEmailSender, EmailSender>();
+            if (options == null)
```
Comment (DmyMi) [inline OutOfSchool/OutOfSchool.EmailSender/ServiceProviderExtensions.cs:13]: remember we were talking about an ability to disable e-mail config for local testing? Maybe add an implementation to the sender service method `Task SendAsync(Message message)` that if the config is missing or some option is specified to disable sending the message, then it just prints it to `log.Debug` so we can copy the link or verify email content locally.

File: OutOfSchool/OutOfSchool.EmailSender/EmailSender.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+using MimeKit;
+using MimeKit.Text;
+
+namespace OutOfSchool.EmailSender
+{
+    public class EmailSender : IEmailSender
+    {
+        IOptions<EmailOptions> _emailOptions;
+        IOptions<SmtpOptions> _smtpOptions;
+
```
Comment (yfedo) [inline OutOfSchool/OutOfSchool.EmailSender/EmailSender.cs:13]: fields can be readonly

File: OutOfSchool/OutOfSchool.EmailSender/EmailSender.cs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            mimeMessage.From.Add(new MailboxAddress(message.From.Name, message.From.Address));
+            mimeMessage.Subject = message.Subject;
+            mimeMessage.Body = new TextPart(TextFormat.Html)
+            {
+                Text = message.Content
+            };
+            return mimeMessage;
+        }
+
+        private async Task SendAsync(MimeMessage mimeMessage)
```
Comment (yfedo) [inline OutOfSchool/OutOfSchool.EmailSender/EmailSender.cs:59]: better to renamr this as SendMimeMessageAsync

File: OutOfSchool/OutOfSchool.EmailSender/EmailSender.cs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        IOptions<SmtpOptions> _smtpOptions;
+
+        public EmailSender(IOptions<EmailOptions> emailOptions, IOptions<SmtpOptions> smtpOptions)
+        {
+            _emailOptions = emailOptions;
+            _smtpOptions = smtpOptions;
+        }
+
+        public Task SendAsync(Message message)
+        {
```
Comment (yfedo) [inline OutOfSchool/OutOfSchool.EmailSender/EmailSender.cs:21]: currently this method is used only in SendAsync(string email, string subject, string htmlMessage)
it lokks like in can be removed from the IEmailSender interface and converted to private method

File: OutOfSchool/OutOfSchool.EmailSender/EmailOptions.cs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+﻿namespace OutOfSchool.EmailSender
+{
+    public class EmailOptions
+    {
+        public static string SectionName => "Email";
+
```
Comment (yfedo) [inline OutOfSchool/OutOfSchool.EmailSender/EmailOptions.cs:6]: inconsistent whitespace))

File: OutOfSchool/OutOfSchool.EmailSender/IEmailSender.cs [hunk 7]
```diff
@@ -0,0 +1,11 @@
+﻿using System.Threading.Tasks;
+
+namespace OutOfSchool.EmailSender
+{
+    public interface IEmailSender
+    {
+        Task SendAsync(Message message);
+
```
Comment (yfedo) [inline OutOfSchool/OutOfSchool.EmailSender/IEmailSender.cs:8]: this method is only called on class where it is implemented

File: OutOfSchool/OutOfSchool.EmailSender/ServiceProviderExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+        public static IServiceCollection AddEmailSender(
+            this IServiceCollection services,
+            Action<OptionsBuilder<EmailOptions>> emailOptions,
+            Action<OptionsBuilder<SmtpOptions>> smtpOptions)
+        {
+            services.AddSingleton<IEmailSender, EmailSender>();
+            if (emailOptions == null)
+            {
+                throw new ArgumentNullException(nameof(emailOptions), @"Please provide emailOptions for EmailSender");
+            }
```
Comment (yfedo) [inline OutOfSchool/OutOfSchool.EmailSender/ServiceProviderExtensions.cs:18]: strings can be moved to resources

File: OutOfSchool/OutOfSchool.IdentityServer/ViewModels/ResetPasswordViewModel.cs [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    {
+        [DataType(DataType.EmailAddress)]
+        [Required(ErrorMessage = "Email is required")]
+        public string Email { get; set; }
+
+        [Required(ErrorMessage = "Password is required")]
+        [RegularExpression(
+            @"^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,}$",
+            ErrorMessage = "Password must contain at least one capital, number and symbol(@$!%*?&).")]
+        [DataType(DataType.Password)]
```
Comment (yfedo) [inline OutOfSchool/OutOfSchool.IdentityServer/ViewModels/ResetPasswordViewModel.cs:15]: this regex can be moved to some CommonConstants or simillar class

**Review by yfedo — APPROVED** [review]

File: OutOfSchool/OutOfSchool.EmailSender/EmailSender.cs (hunk 3 above)
Comment (SergeyNovitsky) [reply OutOfSchool/OutOfSchool.EmailSender/EmailSender.cs:13]: made it readonly and private

File: OutOfSchool/OutOfSchool.EmailSender/EmailSender.cs (hunk 4 above)
Comment (SergeyNovitsky) [reply OutOfSchool/OutOfSchool.EmailSender/EmailSender.cs:59]: renamed

File: OutOfSchool/OutOfSchool.EmailSender/EmailSender.cs (hunk 5 above)
Comment (SergeyNovitsky) [reply OutOfSchool/OutOfSchool.EmailSender/EmailSender.cs:21]: removed from interface and made it private

File: OutOfSchool/OutOfSchool.EmailSender/EmailOptions.cs (hunk 6 above)
Comment (SergeyNovitsky) [reply OutOfSchool/OutOfSchool.EmailSender/EmailOptions.cs:6]: fixed and also fixed the same in SmtpOptions

File: OutOfSchool/OutOfSchool.EmailSender/IEmailSender.cs (hunk 7 above)
Comment (SergeyNovitsky) [reply OutOfSchool/OutOfSchool.EmailSender/IEmailSender.cs:8]: removed from interface

File: OutOfSchool/OutOfSchool.EmailSender/ServiceProviderExtensions.cs (hunk 8 above)
Comment (SergeyNovitsky) [reply OutOfSchool/OutOfSchool.EmailSender/ServiceProviderExtensions.cs:18]: removed second parameter because it repeats standard message

**Review by DmyMi — APPROVED** [review]

**Pull request merged**